src/java/org/apache/cassandra/gms/Gossiper.java

```diff
@@ -211,59 +193,123 @@ public void clearUnsafe()
         seedsInShadowRound.clear();
     }
 
-    // returns true when the node does not know the existence of other nodes.
-    private static boolean isLoneNode(Map<InetAddressAndPort, EndpointState> epStates)
-    {
-        return epStates.isEmpty() || epStates.keySet().equals(Collections.singleton(FBUtilities.getBroadcastAddressAndPort()));
-    }
-
-    final Supplier<ExpiringMemoizingSupplier.ReturnValue<CassandraVersion>> upgradeFromVersionSupplier = () ->
+    private class DefaultClusterVersionProvider implements IClusterVersionProvider
     {
-        // Once there are no prior version nodes we don't need to keep rechecking
-        if (!upgradeInProgressPossible)
-            return new ExpiringMemoizingSupplier.Memoized<>(null);
-
-        CassandraVersion minVersion = SystemKeyspace.CURRENT_VERSION;
+        // -1L means that the cluster may be in upgrading state; positive value is the timestamp of the cluster
+        // was detected as fully upgraded
+        private final AtomicLong notUpgradingSince = new AtomicLong(-1L);
 
-        // Skip the round if the gossiper has not started yet
-        // Otherwise, upgradeInProgressPossible can be set to false wrongly.
-        // If we don't know any epstate we don't know anything about the cluster.
-        // If we only know about ourselves, we can assume that version is CURRENT_VERSION
-        if (!isEnabled() || isLoneNode(endpointStateMap))
+        // returns true when the node does not know the existence of other nodes
+        private boolean isLoneNode(Map<InetAddressAndPort, EndpointState> epStates)
         {
-            return new ExpiringMemoizingSupplier.NotMemoized<>(minVersion);
+            return epStates.isEmpty() || epStates.keySet().equals(Collections.singleton(FBUtilities.getBroadcastAddressAndPort()));
         }
 
-        // Check the release version of all the peers it heard of. Not necessary the peer that it has/had contacted with.
-        boolean allHostsHaveKnownVersion = true;
-        for (InetAddressAndPort host : endpointStateMap.keySet())
+        // minimum time that needs to pass after the cluster is detected as fully upgraded
+        // to report that there is no upgrade in progress
+        private final long MIN_STABLE_DURATION_MS = Long.getLong("cassandra.cluster_version_provider.min_stable_duration_ms", 60_000);
```

Comment:
Why do we need this period? It seems like an addition not strictly needed for this refactor.

Reply:
This is not strictly for CNDB, but rather for HCD. 

tl;dr 
In a situation when we cannot see nodes other than the nodes with the current version, it can be the upgrade or stable state. We don't know if something is going to join the cluster or not. So, this is a safety period. It was working like this before however is less cleaner way I think; Also, previously there was a bug which made it report the cluster was being upgraded forever when it was a single-node cluster, which is fixed now.

Reply:
60s sounds too low to know "noone's gonna join". suppose there's one last node to upgrade. It gets restarted for whatever reason, and then does some long running task on startup (reindex? pulling data?).

Reply:
I think you have in a mind a bit different scenario than this property is for. 

In all scenarios where we have known nodes that are part of the cluster (they were not explicitly removed), which can be on either previous or unknown version (because they are down), we don't claim that cluster is stable state.

This expiration time is for the situation when we know all the nodes are up and at the current version. Internally (inside the `DefaultClusterVersionProvider`) we set the flag that the cluster is stable immediately, but externally (in `Gossiper` method) we consider the cluster stable if it is stable for at least that amount of time. SO in other words, it is for the situation we may want to join a *new* node to the cluster or the information about some nodes is lost for various reasons.

Obviously, in certain situations it makes no sense, especially if we know what we are doing. In other situations this period should be longer. It should be a good default for an ordinary user of CC as an open source DB to run it locally for experimenting. Perhaps the default value should be even shorter. Especially for testing we could just make it zero.